lib/_http_parser.js

```diff
@@ -0,0 +1,890 @@
+var FLAG_UPGRADE = 1 << 5;
+var FLAG_SKIPBODY = 1 << 6;
+var FLAG_SHOULD_KEEP_ALIVE = 1 << 7;
+var FLAG_ANY_UPGRADE = FLAG_UPGRADE | FLAG_CONNECTION_UPGRADE;
+
+function HTTPParser(type) {
+  this.onHeaders = null;
+  this.onBody = null;
+  this.onComplete = null;
+
+  // extra stuff tagged onto parser object by core modules/files
+  this.onIncoming = null;
+  this.incoming = null;
+  this.socket = null;
+
+  this.reinitialize(type);
+}
+HTTPParser.prototype.reinitialize = function(type) {
+  this.execute = this._executeStartLine;
+  this.type = type;
+  if (type === REQUEST)
+    this._state = STATE_REQ_LINE;
+  else
+    this._state = STATE_STATUS_LINE;
+
+  this._err = null;
+  this._flags = 0;
+  this._contentLen = null;
+  this._nbytes = 0;
+  this._nhdrbytes = 0;
+  this._nhdrpairs = 0;
+  this._buf = '';
+  this._seenCR = false;
+
+  // common properties
+  this.headers = [];
+  this.httpMajor = 1;
+  this.httpMinor = null;
+  this.maxHeaderPairs = 2000;
+
+  // request properties
+  this.method = null;
+  this.url = null;
+
+  // response properties
+  this.statusCode = null;
+  this.statusText = null;
+};
+HTTPParser.prototype.finish = function() {
+  var state = this._state;
+  if (state === STATE_BODY_EOF) {
+    this.execute = this._executeBodyIgnore;
+    this._state = STATE_COMPLETE;
+    this.onComplete && this.onComplete();
+  } else if (this.execute !== this._executeError &&
+             state !== STATE_REQ_LINE &&
+             state !== STATE_STATUS_LINE &&
+             state !== STATE_COMPLETE) {
+    return this._setError('Invalid EOF state');
+  }
+};
+HTTPParser.prototype.close = function() {};
+HTTPParser.prototype.pause = function() {};
+HTTPParser.prototype.resume = function() {};
+HTTPParser.prototype._setError = function(msg) {
+  var err = new Error(msg);
+  this.execute = this._executeError;
+  this._err = err;
+  return err;
+};
+HTTPParser.prototype._processHdrLine = function(line) {
+  switch (this._state) {
+    case STATE_HEADER:
+      if (line.length === 0) {
+        // We saw a double CRLF
+        this._headersEnd();
+        return;
+      }
+      var headers = this.headers;
+      var headerslen = headers.length;
+      var fieldName;
+      var fieldValue;
+      var m = RE_HEADER.exec(line);
+      if (m === null) {
+        var firstChr = line.charCodeAt(0);
+        if (firstChr !== 32 & firstChr !== 9)
+          return this._setError('Malformed header line');
+        // RFC 7230 compliant, but less backwards compatible:
+        var extra = ltrim(line);
+        if (extra.length > 0) {
+          if (headerslen === 0)
+            return this._setError('Malformed header line');
+          fieldName = headers[headerslen - 2];
+          fieldValue = headers[headerslen - 1] + ' ' + extra;
+          // Need to re-check value since matched values may now exist ...
+          if (equalsLower(fieldName, CC_CONNECTION)) {
+            if (fieldValue.search(RE_CONN_CLOSE) > -1)
```

Comment:
`regexp.test()` is actually a tad slower believe it or not. It's a nano optimization.